Jetpack Backup: Prevent restores with invalid creds

Initsogar opened the pull request:
We require to prevent site restores if Jetpack Credentials are no longer valid.
We implemented a way to validate credentials in #66468 and #67589 on Calypso Blue and Calypso Green settings page. The idea is to reuse it and disable restore buttons as soon as we detect they are invalid.

#### Proposed changes
- [x] Disable `Restore to this point` button in actions toolbar on Activity Log when credentials are invalid
- [x] Disable `Restore to this point` button in main backups page when credentials are invalid
- [x] Show banner requesting credentials if credentials are invalid
- [x] Introduce new `QueryJetpackCredentialsStatus` data component that will trigger actions for testing Jetpack Credentials to determinate whether they are still valid or not.
- [x] Trigger credentials test validation in backup page
- [x] Trigger credentials test validation in Activity List component

#### Reducer changes
- [x] Jetpack Credentials `testRequestStatus` reducer is now used for handling the test credentials request status, so we can use it to validate is there is an ongoing request to avoid unnecessary API calls.
- [x] Jetpack Credentials `testStatus` reducer is now used for handling the status of the credentials. It will handle `pending`, `valid` and `invalid` statuses.
- [x] Refactored `getJetpackCredentialsTestStatus` and unit tests to consider the changes in reducers.

#### Screenshots

We are reusing the same messages and behaviors that we have when the customer has no credentials.

Component | Valid credentials | Invalid credentials
--- | --- | ---
Backup page (Banner and Daily Backup) | ![image](https://user-images.githubusercontent.com/1488641/194476053-90efa5eb-d2eb-4e4c-8586-9c141945abc0.png) | ![image](https://user-images.githubusercontent.com/1488641/194476175-4e1914b3-7839-47ca-8a4c-93a5602ebd28.png)
Actions toolbar in Activity Log item (Backup and Activity Log pages) | ![image](https://user-images.githubusercontent.com/1488641/194476263-e8f0a6a8-fd81-47d9-89d4-4588c4c57e57.png) |  ![image](https://user-images.githubusercontent.com/1488641/194476326-e06ae093-c02d-4500-b418-74fb58afde4b.png)

#### Testing Instructions
<!--
Add as many details as possible to help others reproduce the issue and test the fix.
“Before / After” screenshots can also be very helpful when the change is visual.
-->

* **Jetpack Cloud / Calypso Green**
  * Grab a site with a Jetpack Backup plan. Preferable, a site you can modify the SSH/SFTP password.
  * Ensure you have working SSH/SFTP credentials on `Settings` page.
  * Navigate to `Activity Log` and `Backup` sections and ensure you can see the `Restore to this point` buttons in green color (that means they are enabled).
    * Also you should see the banner with the message _“Every change you make will be backed up”_ on top of Backup page.
  * Modify the SSH/SFTP password on your site server (not on Jetpack Cloud settings).
  * Navigate again to `Activity Log` or `Backup` page and validate the same buttons are now disabled. It may take a couple of seconds as it is trying to test credentials using an API.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the “[Status] String Freeze” label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the “fixes” keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

#### References
* peaFOp-7G-p2
* 1202895132770065-as-1202906937479000

Initsogar committed:
Add getAreJetpackCredentialsInvalid selector

This selector is useful to validate if existing Jetpack Credentials are still valid. Also including unit tests for different scenarios.

Initsogar committed:
Remove default assertions

Initsogar committed:
Disable restore buttons when creds are invalid

This disable restore buttons on main backup page and action toolbar in activity log when credentials are invalid.

Initsogar committed:
Display enable restore banner if creds are invalid

Initsogar committed:
Fix credentials invalid selector imports

Initsogar committed:
Add new selector for handling creds test requests

Also refactoring current selector and unit tests related.

Initsogar committed:
Fix request creds selector to default false value

If there’s no value on testRequestStatus yet for specific siteId and role, we should return `false`.

Initsogar committed:
Refactor unit tests for creds test selectors

Initsogar committed:
Add QueryJetpackCredentialsStatus data component

Initsogar committed:
Fix optional chaining to fetch jetpack creds state

Initsogar committed:
Display credentials warning text when are invalid

Initsogar committed:
Add test credentials status on backup page

Initsogar committed:
Add test credentials status on activity card list

Initsogar committed:
Add creds test and disable restores in ActivityLog

This change include the option to test credentials on ActivityLog that is currently being used on Calypso Blue, and we are now disabling restore buttons here as well.

mavegaf reviewed (approved):
I tested on Calypso Blue and Green, and it worked in both cases. 

For my site, it took around 12 seconds to finish the credentials tests before disabling the restore button, but I think it is ok for this iteration. We first can check the impact of the current change and later decide if it should be improved.

tyxla commented on client/components/activity-card-list/index.jsx:
```diff
@@ -321,6 +322,7 @@ class ActivityCardList extends Component {
 				<QueryRewindPolicies siteId={ siteId } />
 				<QueryRewindCapabilities siteId={ siteId } />
 				<QueryRewindState siteId={ siteId } />
+				<QueryJetpackCredentialsStatus siteId={ siteId } role="main" />
```
Should this be rendered in the component where we actually use the selectors to render the credentials?

Initsogar replied:
> Should this be rendered in the component where we actually use the selectors to render the credentials?

The components that use the `areJetpackCredentialsInvalid()` selector are action buttons, and we use them to decide whether we need to disable buttons. Given there may be multiple action buttons in the same page, we considered a better option to render this `QueryJetpackCredentialsStatus` in a parent component once.

tyxla commented on client/components/data/query-jetpack-credentials-status/README.md:
```diff
@@ -0,0 +1,37 @@
+# Query Jetpack Credentials Status
+
+`<QueryJetpackCredentialsStatus />` is a React component which dispatches actions for testing Jetpack Credentials to determinate if they are still valid or not.
```
Left a bit of suggestions for improvement here:

```suggestion
`<QueryJetpackCredentialsStatus />` is a React component that dispatches actions for testing Jetpack Credentials to determine whether they are still valid.
```

Initsogar replied:
Sounds reasonable, thanks for that one!

tyxla commented on client/state/jetpack/credentials/reducer.js:
```diff
@@ -71,9 +71,26 @@ export const testRequestStatus = keyedReducer( 'siteId', ( state, { type, role }
+export const testStatus = keyedReducer( 'siteId', ( state, { type, role } ) => {
+	switch ( type ) {
+		// Do not affect current state if executing a new test
+		case JETPACK_CREDENTIALS_TEST:
```
This should not be necessary, since the `type` will not match any of the existing `case`s. We should be safe to remove it.

Initsogar replied:
Yeah, you are right. I added it while testing because I did forget to `return $state` at the end, then I added it back but never fixed this. Working on it :)

tyxla commented on client/state/jetpack/credentials/selectors.js:
```diff
@@ -0,0 +1,9 @@
+import 'calypso/state/jetpack/init';
+
+export function getAreJetpackCredentialsInvalid( state, siteId, role ) {
+	return 'invalid' === state.jetpack?.credentials?.testStatus?.[ siteId ]?.[ role ];
```
Since we already import the `jetpack` reducer on L1, `state.jetpack.credentials.testStatus` will always exist, so we don't need all the optional chaining here:

```suggestion
	return 'invalid' === state.jetpack.credentials.testStatus[ siteId ]?.[ role ];
```

Initsogar replied:
Fixed 👍

tyxla commented on client/state/jetpack/credentials/selectors.js:
```diff
@@ -0,0 +1,9 @@
+}
+
+export function isRequestingJetpackCredentialsTest( state, siteId, role ) {
+	return state.jetpack?.credentials?.testRequestStatus?.[ siteId ]?.[ role ] || false;
```
Same as above:
```suggestion
	return state.jetpack.credentials.testRequestStatus[ siteId ]?.[ role ] || false;
```

Initsogar replied:
Fixed 👍

tyxla commented on client/state/jetpack/credentials/test/reducer.js:
```diff
@@ -1,61 +1,123 @@
+			} );
+		} );
+
+		test( 'should keep existing roles request statuses after requesting a new credentials test for a new role', () => {
```
Do we also need a test to ensure that we store data for multiple sites at the same time, and data of a new site doesn't overwrite data of previously existing site?

Initsogar replied:
Yes, great suggestion. Just added one test case for that!

tyxla commented on client/state/jetpack/credentials/test/reducer.js:
```diff
@@ -1,61 +1,123 @@
+			} );
+		} );
+
+		test( 'should keep previous value if we started a new credentials test for same role', () => {
```
Just as above, do we need a test that demonstrates that data is properly stored and maintained for multiple sites at the same time?

Initsogar replied:
Thanks! Just added two test case for that :)

tyxla commented on client/state/jetpack/credentials/selectors.js:
```diff
@@ -0,0 +1,9 @@
+import 'calypso/state/jetpack/init';
+
+export function getAreJetpackCredentialsInvalid( state, siteId, role ) {
```
I'd rename this to:

```suggestion
export function areJetpackCredentialsInvalid( state, siteId, role ) {
```

because `getAre` feels counter-intuitive. WDYT?

Initsogar replied:
Thanks for this! I'm renaming it.

tyxla reviewed (commented):
This looks nice overall 👍 I haven't tested, but have a few suggestions on the code side of things.

Initsogar committed:
Update client/components/data/query-jetpack-credentials-status/README.md

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

Initsogar committed:
Remove unnecessary case condition

Initsogar committed:
Rename areJetpackCredentialsInvalid selector

This commit renames getAreJetpackCredentialsInvalid to areJetpackCredentialsInvalid on all instances, including unit tests.

It also removes unnecessary optional chaining operators on Jetpack Credentials selectors.

Initsogar committed:
Add test cases to support multiple sites data

Initsogar commented:
@tyxla Thanks for all your suggestion! I made the changes suggested, including the unit tests recommendations. It would be great if you can re-review it :)

Initsogar committed:
Merge remote-tracking branch 'origin/trunk' into add/prevent-restore-invalid-creds

tyxla reviewed (approved):
LGTM from the code perspective 🚀
